### PR TITLE
ADR proposal: AWS Backup

### PR DESCRIPTION
**Pull request by s0enke**

### CODE REVIEW

**Comment by yawn** [comment]
Meta:

1. Typo in "Integrity"

File: docs/adrs/0001-aws-backup.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+
+# Backups with AWS Backup
+
+## Status
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:4]: I'd vote for implicit status: everything that's non main/master is proposed, everything on main/master is implemented until deleted.

File: docs/adrs/0001-aws-backup.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+proposed
+
+### TODO
+
+ - [ ] Spike solution
+ - [ ] Spike AWS Config vs. Tag Policies
+ - [ ] Decide whether opt-out mechanism is
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:12]: WDYM?

File: docs/adrs/0001-aws-backup.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+ - [ ] Spike solution
+ - [ ] Spike AWS Config vs. Tag Policies
+ - [ ] Decide whether opt-out mechanism is 
+
+## Context
+
+A well-architected AWS setup includes backups for non-transient resources like RDS/Aurora, DynamoDB, EFS etc. so superwerker should protect users by enabling backups automatically.
+Since superwerker prefers the usage of native AWS services, AWS Backup is used.
+AWS Backup does not currently support backing up all resources in an AWS account. Either ARNs or Tags have to be speficied. A workaround has to be found.
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:18]: Typo. Semantically: we could think about implementing an inheritance mechanism for accounts and OUs (hierarchically) that initially only AWS Backup supported resource types. Drawback: this Lambda most likely needs either CT CAR support or depends on a shared infra pipeline - I'd think CAR support would be the easiest in the beginning (deployment is master/payer triggered anyways) and the integrity protection for the tag prefix could be easily deployed on the top level as well.

File: docs/adrs/0001-aws-backup.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Decision
+
+- Use AWS Organizations Backup Policies to enforce Backups across the entire Organization
+- Backups of resources are created daily
+- The default Backup Vault is used
+- Snapshots are protected via Integrity Protection SCP, only AWS Backup role can delete Snapshots 
+- Enforce Tags, either
+  - Use AWS Config Rule `required-tags` to check if tags have been set on backup-eligible resources, set tags via AWS Confug Rules Remediation
+  - or AWS Organizations Tag Policies and CW Events to tag resources without backup tag
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:28]: 👍

File: docs/adrs/0001-aws-backup.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+- Tag name is `superwerker:backup`, only valid value for now is `daily`
+- No cross-region backup right now to keep it simple.
+
+## Consequences
+
+- AWS Backup is set as the preferred backup solution
+- Resources are backed up by default
+- Opt-out per Resource or per Account/OU can be added later
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:36]: If we go via an inheritance mechanism an opt-out tag might "block" the inheritance on the resource e.g. through a filter where supported.

**Comment by s0enke** [comment]
> Meta:
> 
> 1. Typo in "Integrity"

fixed in issue template

File: docs/adrs/0001-aws-backup.md (hunk 5 above)
Comment (s0enke) [reply docs/adrs/0001-aws-backup.md:35]: I proposed to allow to set the `superwerker:backup` tag to `none` to opt-out of backups per resource. That's good enough for 1.0 imho. What do you think?

File: docs/adrs/0001-aws-backup.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+
+# Backups with AWS Backup
+
+### TODO
+
+ - [ ] Spike solution
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:6]: Done

File: docs/adrs/0001-aws-backup.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ - [ ] Spike AWS Config vs. Tag Policies
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:7]: Done

Comment (yawn) [reply docs/adrs/0001-aws-backup.md:7]: And document!

File: docs/adrs/0001-aws-backup.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ - [ ] Decide whether to backup EC2 instances, document decision
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:8]: Done & document. Answer is none, only EBS.

File: docs/adrs/0001-aws-backup.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ - [ ] Decide whether enabling/disabling particular resource types ("Service Opt-in") by the user should be allowed 
+
+## Context
+
+A well-architected AWS setup includes backups for non-transient resources like RDS/Aurora, DynamoDB, EFS etc. so superwerker should protect users by enabling backups automatically.
+Since superwerker prefers the usage of native AWS services, AWS Backup is used.
+AWS Backup does not currently support backing up all resources in an AWS account. Either ARNs or Tags have to be specified. A workaround has to be found.
+
+So we need a way to tag all resources which should be backed up automatically. Tag policies come into mind. But Tag Policies do not enforce tags on untagged resources ([docs](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_tag-policies-enforcement.html)).
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:17]: tag policies

File: docs/adrs/0001-aws-backup.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+## Decision
+
+- Use AWS Organizations Backup Policies to enforce Backups across the entire Organization
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:21]: backup policies / backups / organization

File: docs/adrs/0001-aws-backup.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+- Backups of resources are created daily
+- The default Backup Vault is used
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:23]: backup vault

File: docs/adrs/0001-aws-backup.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+- Snapshots are protected via Integrity Protection SCP, only AWS Backup role can delete Snapshots 
+- Enforce Tags, either
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:25]: tags

File: docs/adrs/0001-aws-backup.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  - Use AWS Config Rule `required-tags` to check if tags have been set on backup-eligible resources, set tag to `daily` if tag is not `daily` or `none` via AWS Config Rules Remediation
+  - or AWS Organizations Tag Policies and CW Events to tag resources without backup tag
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:27]: update decision to actual state

File: docs/adrs/0001-aws-backup.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+- Tag name is `superwerker:backup`, only valid value for now is `daily` or `none`, default is `none`
+- No cross-region backup right now to keep it simple.
+- Opt-out is possible via `supwerker:backup` tag set to `none`
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:30]: `superwerker:backup`

File: docs/adrs/0001-aws-backup.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+### AWS Config Organizational Rules vs. Conformance Packs vs. StackSets
+
+There are several ways to deploy config rules org-wide.
+
+ - AWS Config Organizational Rules and Conformance Packs have no fault-tolerance. If one Sub-Account does not have a Config Recorder installed, the entire installation fails.
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:36]: Depends on #57

File: docs/adrs/0001-aws-backup.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+ - Conformance packs need a S3 bucket, which makes the setup more complex
```
Comment (yawn) [inline docs/adrs/0001-aws-backup.md:37]: Basically requires a stackset with inline instance (target the log account). We also need to deploy remediation roles for all the accounts and the service role (not the service linked role).

**Review by yawn — CHANGES_REQUESTED** [review]
LGTM*

**Comment by s0enke** [comment]
superseded by #81

**Pull request closed without merging**